src/execution/value.rs

```diff
@@ -261,14 +261,141 @@ pub enum Value {
     F64(F64),
     // F64,
     // V128,
+    Ref(Ref),
 }
 
-#[derive(Clone, Debug, PartialEq)]
-#[allow(dead_code)]
+#[derive(Clone, Copy, Debug, PartialEq)]
 pub enum Ref {
-    Null,
-    // Func,
-    // Extern,
+    Func(FuncAddr),
+    Extern(ExternAddr),
+}
+
+impl Ref {
+    pub fn default_from_ref_type(rref: RefType) -> Self {
+        match rref {
+            RefType::ExternRef => Self::Extern(ExternAddr::default()),
+            RefType::FuncRef => Self::Func(FuncAddr::default()),
+        }
+    }
+
+    pub fn is_null(&self) -> bool {
+        match self {
+            Self::Extern(extern_addr) => extern_addr.is_null,
+            Self::Func(func_addr) => func_addr.is_null,
+        }
+    }
+
+    pub fn is_specific_func(&self, func_id: u32) -> bool {
+        match self {
+            Self::Func(func_addr) => !func_addr.is_null && func_addr.addr == func_id as usize,
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl Display for Ref {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Ref::Func(func_addr) => write!(f, "FuncRef({})", func_addr),
+            Ref::Extern(extern_addr) => write!(f, "ExternRef({})", extern_addr),
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub struct FuncAddr {
+    pub is_null: bool,
+    // it is the idx of the function in the current module
+    pub addr: usize,
+}
+
```

Comment:
I know we discussed this in depth, but I think I might have changed my mind. Sorry :(

Okay, this isn't a blocking issue for me, but I would like to bring to the discussion making this a newtype
```rs
pub struct FuncAddr(Option<usize>)
```

This would add more "type safety", in the sense that a) a FuncAddress with a non-zero `addr` and true `is_null` shouldn't exist as a state in the code, and b) to avoid situations in which we forget to check `is_null`, the option variant is more safe

Reply:
Again, please feel free to push back on this. "It already works" is a valid argument. I'd just like to send it up for discussion

Reply:
Let's go for the Option choice. Might be a lot of rework, but might also be much easier to understand.

Reply:
Use just a USIZE
If u32::MAX => NULL
Do it like wasmtime B)

Reply:
Let's check the spec on this

Reply:
with the Option!!!!! 🤔